### ISSUE TITLE
The "sync" command needs "rcon-cli" added to it

### ISSUE THREAD

**Issue by dreylok** [itzg/docker-mc-backup#189]
The following line in backup-loop.sh is missing the "rcon-cli" portion of the command:
`retry ${RCON_RETRIES} ${RCON_RETRY_INTERVAL} sync`

This results in the command being executed in the mc-backup container instead, calling the Linux "sync" command. There is no need for this command to run inside the mc-backup container and it appears to have been intended to be the `sync` command in Minecraft server, which should follow the save-all command.

Suggested fix is to simply add the "rcon-cli" command to this line. Will submit a PR; opening this issue for tracking purposes.

**Comment by itzg** [comment]
I looked later and there is no vanilla sync command

https://minecraft.wiki/w/Commands

I think the original author really meant the Linux sync command. Now how important that was is a good question. I looked and I must have missed the change during PR and didn't ask.

**Comment by dreylok** [comment]
I stand corrected. I swear I found something obscure that indicated there was a sync command for Minecraft server and even double checked in ChatGPT (for what that's worth) and it even went so far as to elaborate the difference between "sync" and "save-all flush" commands. But you're right, there is no reference to it and using the command in Minecraft server returns an error.

With that said, I agree the original author intended to use the Linux command, but it somehow got wrapped in the rcon-cli retry loop erroneously. However, running the sync command in the mc-backup container would not be effective since any potential data would be resident in the memory of the minecraft-server container.

Docker should be handling this at the volume level itself from what I understand since it's intended that data can be shared between services, but if we wanted to be absolutely certain, the command would need to be executed in the server container. Is there a way to facilitate this from the backup container?

**Comment by itzg** [comment]
> Docker should be handling this at the volume level itself from what I understand since it's intended that data can be shared between services, but if we wanted to be absolutely certain, the command would need to be executed in the server container. Is there a way to facilitate this from the backup container?

I don't believe Docker would be doing anything special since it's all just Linux inodes mounted into the container/process. Even being called from an adjacent container, it's likely it works as intended since again it's the same filesystem/inode mounted into both containers.

...however, if we're doubting its effectiveness and why it's even there, then probably should just remove the call entirely.

**Comment by dreylok** [comment]
From what I'm reading, the Docker volume "driver" does consistent writes by default to ensure data is immediately available on the host, and assumedly the other containers accessing named volumes. Nothing to worry about here.

However, the OS in the minecraft-server container could potentially have data in cache that has not yet been submitted to storage. Docker cannot do anything about that, and that's where the sync command could help ensure data is truly flushed to the storage layer. If we're erring on the side of caution, perhaps it should be included?

**Comment by itzg** [comment]
> From what I'm reading, the Docker volume "driver" does consistent writes by default to ensure data is immediately available on the host, and assumedly the other containers accessing named volumes. Nothing to worry about here.

That'll be for Docker Desktop, right? I was thinking about fully native Docker on Linux on where everything (host and containers) are running in the same kernel against the same filesystems and inodes.

> However, the OS in the minecraft-server container could potentially have data in cache that has not yet been submitted to storage. Docker cannot do anything about that, and that's where the sync command could help ensure data is truly flushed to the storage layer. If we're erring on the side of caution, perhaps it should be included?

But yes, a point I forgot I was going to make is that it has been working fine for 5 years, so no need to mess with it:

https://github.com/itzg/docker-mc-backup/commit/ce2ea1327e2b515313652c39068566339872b7d4#diff-26c954788cf358925d71c7e1ef6dd3824e2884ee29afe73d58091f0775cc5611R286

In conclusion :)
- it's not a typo
- it was intended to serve a purpose
- it's possibly fulfilling that purpose in some cases (fully native Linux)
- even if/when it doesn't do any useful operation, it apparently doesn't hurt anything since it's been there for ages

So, just close the issue?

**Comment by Krymonota** [comment]
I believe that the `sync` call causes intermittent TPS drops and brief save lag across multiple Minecraft pods on the same node in my Kubernetes cluster.

While the preceding `rcon-cli save-all flush` is the correct and sufficient way to persist world data to disk from the server side, the subsequent `sync` command initiates a system-wide kernel flush of all dirty pages - not just for the container or its volume, but for the entire node's file system.

This results in short bursts of intense `jbd2` (ext4 journaling) activity, which blocks concurrent writes and momentarily stalls unrelated pods. In high-density environments, this can degrade performance across multiple game servers.

I've opened #231 to disable the `save-all flush` step for our use case, but I'd also suggest making the `sync` call optional or disabled by default, as it's generally unnecessary when the game server is already handling data consistency via `flush`.

**Comment by Krymonota** [comment]
Just to expand on the `save-all flush` + `sync` interaction:

While `save-all flush` ensures that world data is written out from the Minecraft server, these writes typically land in the Linux kernel's page cache as dirty pages. Unless the application explicitly calls `fsync()`, the data is buffered - not yet committed to physical storage.

The `sync` command then forces all dirty pages across the entire system to be flushed to disk. When multiple Minecraft pods issue backups around the same time, this causes a synchronized flush storm, which overwhelms the underlying filesystem - particularly on ext4 - leading to high `jbd2` activity and stalled writes across containers.

This effect is especially noticeable on nodes with consumer-grade SSDs or higher pod density, which happens to be the case for my affected node.

So even though `sync` may seem like a safety net (and I'm pretty sure that was the original contributor's intention), in modern containerized environments it often does more harm than good.
Sure, most people probably don't run a dozen backup containers on a single physical machine - but I figured I'd share the insight anyway 😄

**Comment by itzg** [comment]
Great point @Krymonota ! I didn't even realize that `sync` was still there. Any concerns you can think for simply removing that step?

I can still make the save-all configurable just to cover all scenarios.

**Comment by Krymonota** [comment]
> Any concerns you can think for simply removing that step?

Ultimately, it probably still contributes to data integrity, especially since the backup export runs immediately afterward. However, this comes at the cost of I/O performance. In the worst case, removing it could result in an incomplete or corrupted backup if dirty pages haven't been flushed to disk in time.

So on second thought, it's probably better to keep the `sync` step, especially since we can't know the full range of user environments for this image. That said, there should be an option to disable it for advanced or containerized setups like mine.

Maybe these two flags would offer a good flexibility:
- `ENABLE_SYNC` (default: `true`)
- `ENABLE_SAVE_ALL` (default: `true`)
